tests/pallas/gpu_pallas_distributed_test.py

```diff
@@ -0,0 +1,249 @@
+      y, _, _ = jax.jit(
+                shard_map.shard_map(
+                  body, mesh,
+                  in_specs=(P('x'), P(None), P(None)),
+                  out_specs=(P('x'), P(None), P(None)),
+                  check_rep=False
+                )
+      )(x, ready_sem, recv_sem)
+      expected = x[8:] if jax.process_index() == 0 else x[:8]
+      np.testing.assert_allclose(y.addressable_shards[0].data, expected)
+
+    launch_multiprocess_test(self, _test_basic_remote_dma, 2)
+
+  def test_barrier_semaphore(self):
+    def _test_barrier_semaphore(self):
+      def kernel(x_ref, barrier_sem_alias, recv_sem_alias,
+                 y_ref, barrier_sem, recv_sem):
+        del barrier_sem_alias, recv_sem_alias
+        my_id = pl.device_id()
+        num_devices = lax.psum(1, 'x')
+        neighbor = lax.rem(my_id + 1, num_devices)
+        pl.semaphore_signal(barrier_sem.at[0], device_id=neighbor,
+                            device_id_type=pl.DeviceIdType.LOGICAL)
+        pl.semaphore_wait(barrier_sem.at[0])
+        pl.async_remote_copy(
+            x_ref, y_ref, None, recv_sem.at[0], device_id=neighbor,
+            device_id_type=pl.DeviceIdType.LOGICAL
+        ).wait()
+
+      num_devices = jax.device_count()
+      x = jnp.arange(num_devices * 8 * 128).reshape((num_devices * 8, 128))
+      barrier_sem = jnp.zeros((1,), dtype=jnp.int32)
+      recv_sem = jnp.zeros((1,), dtype=jnp.int32)
+
+      def body(x, barrier_sem, recv_sem):
+        return pl.pallas_call(
+            kernel,
+            in_specs=[pl.BlockSpec(memory_space=plgpu.GMEM),
+                      pl.BlockSpec(memory_space=plgpu.GMEM),
+                      pl.BlockSpec(memory_space=plgpu.GMEM)],
+            out_specs=[pl.BlockSpec(memory_space=plgpu.GMEM),
+                       pl.BlockSpec(memory_space=plgpu.GMEM),
+                       pl.BlockSpec(memory_space=plgpu.GMEM)],
+            out_shape=[jax.ShapeDtypeStruct((8, 128), jnp.int32),
+                       jax.ShapeDtypeStruct((1,), jnp.int32),
+                       jax.ShapeDtypeStruct((1,), jnp.int32)],
+            input_output_aliases={1: 1, 2: 2},
+            compiler_params=plgpu.GPUCompilerParams(use_custom_barrier=True),
+        )(x, barrier_sem, recv_sem)
+
+      device_mesh = mesh_utils.create_device_mesh(
+          (jax.device_count(),), jax.devices())
+      mesh = jax.sharding.Mesh(device_mesh, ['x'])
+      y, _, _ = jax.jit(
+                shard_map.shard_map(
+                    body, mesh,
+                    in_specs=(P('x'), P(None), P(None)),
+                    out_specs=(P('x'), P(None), P(None)),
+                    check_rep=False
+                )
+      )(x, barrier_sem, recv_sem)
+      expected = x[8:] if jax.process_index() == 0 else x[:8]
+      np.testing.assert_allclose(y.addressable_shards[0].data, expected)
+
+    launch_multiprocess_test(self, _test_barrier_semaphore, 2)
+
+
+class VerificationTest(jtu.JaxTestCase):
```

Comment:
Is there anything different about this code than the one in `tpu_pallas_distributed_test`?

Reply:
No, it should not. Except the way we pass semaphores.